pam: trigger pam_authenticate on login

awly opened the pull request:
This will trigger any "auth" PAM modules configured on the system for
teleport. For example, Duo 2FA prompt on each connection.
The module will be able to interact with the user (e.g. print prompts).

Also, make PAM env var propagation consistent for port forwarding
sessions.

Fixes https://github.com/gravitational/teleport/issues/3929

Commit:
pam: trigger pam_authenticate on login

This will trigger any "auth" PAM modules configured on the system for
teleport. For example, Duo 2FA prompt on each connection.
The module will be able to interact with the user (e.g. print prompts).

Also, make PAM env var propagation consistent for port forwarding
sessions.

Fixes https://github.com/gravitational/teleport/issues/3929

awly commented:
Tested with basic `pam_exec.so` module and [Duo](https://duo.com/docs/duounix#overview)

russjones commented:
Can you add test coverage, we have a PAM module we use for testing that can be configured to test if Teleport calls the authentication function. https://github.com/gravitational/teleport/blob/master/modules/pam_teleport/pam_teleport.c

webvictim commented:
Some weird test stuff happening here:

Jenkins:
```
signal: killed
FAIL	github.com/gravitational/teleport/lib/pam	74.098s
```

It looks like Drone timed out and killed the container too.

webvictim commented:
Looks like a memory leak:

![2020-07-02-141610__2247x495__maim](https://user-images.githubusercontent.com/897821/86390541-a4786700-bc6e-11ea-9db3-1c9e5959e6c2.png)

Commit:
Revamp PAM testing stack

- update PAM policies and module for "auth" step
- use pam_teleport.so from the repo directory instead of guessing
  OS-specific global path
- add tests covering all failure scenarios and generally refactor PAM
  tests

awly commented:
> Looks like a memory leak:

I think it's because we're reading from `stdin` in some cases and in tests it's set to [an infinite nop reader](https://github.com/gravitational/teleport/blob/master/lib/pam/pam_test.go#L118-L123).

Added an `io.LimitReader` around it for sanity.

@russjones refactored the testing setup and added missing test cases.

awly commented:
Need to re-run `make install` in `modules/pam_teleport` on the build box to get things passing again

awly commented:
ok, I fixed the buildbox build steps, but looks like drone uses a cached version of buildbox.
@webvictim any particular reason we don't use `make -C build.assets test` here: https://github.com/gravitational/teleport/blob/master/.drone.yml#L77-L85 ?
That would let drone pick up any new changes to buildbox.

webvictim commented:
The reason is mostly because Jenkins uses `/gopath` while Drone uses `/go`, so the Docker `-v` commands are different. We also use a shared Go build cache in Drone for speed which Jenkins doesn’t do. I can always modify the `Makefile` to hold the logic and just call those targets from Drone - the reason I didn’t initially is that for rapid iteration, it can be nice to be able to change the build code without having to cut new tags and backport fixes to release branches.

With that said, this has actually got me wondering whether the shared build cache could be influencing the test failures in Drone...

awly commented:
Looks like https://github.com/gravitational/teleport/pull/3982 will soon fix the problem.
I'll wait for it to land, thanks @webvictim!

webvictim commented on build.assets/Makefile:
```diff
@@ -220,3 +221,9 @@ release-centos6-fips: bbox-centos6-fips
 	docker run $(DOCKERFLAGS) -i $(NOROOT) $(BBOX) \
 		/usr/bin/make release -e ADDFLAGS="$(ADDFLAGS)" OS=windows
+
+.PHONY:pam
+pam:
+	make -C ../modules/pam_teleport
+	rm pam/* || mkdir -p pam
+	cp ../modules/pam_teleport/pam_teleport.so ../modules/pam_teleport/policy/* pam/
```
As this is the Makefile in `build.assets`, the PAM build commands here should arguably be run inside a Docker container to reduce the need to have `pam-devel` or similar packages installed on the host.

awly replied:
Great idea!
Moved `modules/pam_teleport` to `build.assets/pam` and now it'll compile `pam_teleport.so` as part of buildbox creation.

webvictim commented on lib/pam/pam_test.go:
```diff
@@ -91,19 +92,11 @@ func (s *Suite) TestEcho(c *check.C) {
 // read in the first argument and set it as a PAM environment variable. This
 // test then validates it matches what was set in the policy file.
 func (s *Suite) TestEnvironment(c *check.C) {
-	// Skip this test if the binary was not built with PAM support.
-	if !BuildHasPAM() || !SystemHasPAM() {
-		c.Skip("Skipping test: PAM support not enabled.")
-	}
-
-	local, err := user.Current()
-	c.Assert(err, check.IsNil)
-
```
Are we now assuming that every single binary is built with PAM support, so tests will fail if it's not present? Or is this controlled by the `pam` tag?

awly replied:
This check is performed in `SetUpSuite`.
If the skip happens there, no tests will run.
There's no need to do the same check in every test now.

Commit:
Build pam_teleport.so during buildbox build inside docker

This removes the need for libpam-devel on the host and reliably compiles
pam_teleport.so in our CI pipeline.
As part of this, combine build.assets/pam/ and modules/pam_teleport to
avoid the need to sync them.

benarent commented:
Do Teleport users need to set anything, or is this automatically on with

```
  pam:
     enabled: true
```

awly commented:
@benarent automatically on with that config field, yes.
Users do need to configure a PAM auth policy to get any effects from this.

webvictim commented:
@awly #3982 is merged; I fixed the Dockerfile merge conflicts for you.

Looks like integration tests are failing.

Commit:
Fix integration PAM test cases

Updated expected module output.

benarent commented:
👍 from me. I tested it with Duo and it worked perfectly. New Video Demo: https://share.getcloudapp.com/YEupwAAl

webvictim commented on build.assets/Dockerfile:
```diff
@@ -43,8 +43,9 @@ RUN (curl -L https://github.com/golangci/golangci-lint/releases/download/v1.24.0
+RUN make -C /opt/pam_teleport install
 
-VOLUME ["/go/src/github.com/gravitational/teleport"]
+VOLUME ["/gopath/src/github.com/gravitational/teleport"]
```
This should be `VOLUME ["/go/src/github.com/gravitational/teleport"]` rather than `VOLUME ["/gopath/src/github.com/gravitational/teleport"]` - might be an idea to rebase?

awly replied:
Done

Commit:
Remove merge conflict marker

awly commented:
Need approval from @klizhentas or @russjones on this one
